### PR TITLE
Force push git mirror

### PR DESCRIPTION
**Pull request by scott-vsi**

### CODE REVIEW

File: linux/git_mirror [hunk 1]
```diff
@@ -813,21 +848,181 @@ function git_push_main()
...
+              grep -E "${GIT_MIRROR_PUSH_REFS_REGEX}" | \
+              grep -v -E "${GIT_MIRROR_SKIP_REFS_REGEX}" | \
+              grep -v -e '^refs/remotes' -e '^refs/stash' || :))
+          # xargs tries all args before returning an error code
+          # per xarg's man page:
+          # "The xargs utility exits immediately (without processing any
+          # further input) if ... an invocation of utility exits with a value
+          # of 255", which is not the case if pushing a branch fails
+          # (If refs is empty, the echo succeeds but there is nothing to pass
+          # through the pipe, so the command short-circuits)
```
Comment (andyneff) [inline linux/git_mirror:865]: If you want to use that exit 255 behavior, you could always:

```
echo blah | xargs bash -c '"${GIT}" push "${repo_url}" || exit 255'
```

File: linux/git_mirror [hunk 2]
```diff
@@ -813,21 +848,181 @@ function git_push_main()
...
+          if ! echo ${refs[@]+"${refs[@]}"} | xargs "${GIT}" push "${repo_url}"; then
+            # Force push those refs that failed
+            for ref in "${refs[@]}"; do
+              # If the push was successful, then there is no need to force push
+              # (If the ref is not found, git ls-remote returns the empty string)
+              if [ "$("${GIT}" ls-remote "${repo_url}" "${ref}" | awk '{print $1}')" != \
+                   "$("${GIT}" show-ref --hash "${ref}")" ]; then
+                # No quotes around REGEX
+                if [[ "${ref}" =~ ${GIT_MIRROR_FORCE_PUSH_REFS_REGEX} ]]; then
```
Comment (andyneff) [inline linux/git_mirror:874]:
```suggestion
                if [[ ${ref} =~ ${GIT_MIRROR_FORCE_PUSH_REFS_REGEX} ]]; then
```

File: linux/git_mirror [hunk 3]
```diff
@@ -813,21 +848,181 @@ function git_push_main()
...
+              fi
+            done
+          fi
 
           if git_mirror_has_lfs; then
             next_section "Pushing lfs objects for ${repo_path}"
-            "${GIT}" remote add mirror "${repo_urls[$index]}" 2>/dev/null || "${GIT}" remote set-url mirror "${repo_urls[$index]}"
+            "${GIT}" remote add mirror "${repo_url}" 2>/dev/null || \
+                "${GIT}" remote set-url mirror "${repo_url}"
             # Does not work on file systems, only with real lfs servers, unless you set up lfs-filestore
```
Comment (andyneff) [inline linux/git_mirror:913]: That's only true if you are talking about what you are pushing _to_, not _from_. And that's always true about git lfs

File: linux/git_mirror [hunk 4]
```diff
@@ -813,21 +848,181 @@ function git_push_main()
...
         popd &> /dev/null
       else
-        echo "No dir found for submodule ${repo_paths[$index]}"
+        echo "No dir found for submodule ${repo_path}"
+      fi
+    done
+  popd &> /dev/null
+
+  # Reset the color
+  echo -n $'\e[0m'
```
Comment (andyneff) [inline linux/git_mirror:924]: Why is this necessary?

File: linux/git_mirror [hunk 5]
```diff
@@ -813,21 +848,181 @@ function git_push_main()
...
+            #
+            # git has two kinds of tags: lightweight and annotated. Annotated
+            # tags are commits of their own [1]. In order to list the commit
+            # referred to by the tag (instead of the tag's commit itself) with
+            # git show-ref (from a local repo), use the --dereference flag, or
+            # with git rev-parse (from a local repo) or ls-remote (from a
+            # remote repo), peel the ref with ^{} [2], e.g.,
+            # git rev-pasre refs/tags/atag^{}
+            # We are creating annotated tags (because including a message implies
+            # annotation)
```
Comment (andyneff) [inline linux/git_mirror:1008]: > Annotated tags are meant for release while lightweight tags are meant for private or temporary object labels. For this reason, some git commands for naming objects (like git describe) will ignore lightweight tags by default.

[1] - https://git-scm.com/docs/git-tag

You really shouldn't be misusing annotated tags like that.

File: tests/int/test-git_mirror.bsh [hunk 6]
```diff
@@ -267,13 +288,47 @@ begin_test "Part 5 - Cloning from mirror"
...
+  # I could do these tests in the clone instead (the refs would be listed under
+  # refs/remotes and there would be a separate set of refs/heads; also, there
+  # are cases where certain tags are not fetched without the
+  # git fetch --tags flag)
+  #
+  # git 1.8.3 does not support git -C
+  pushd "${MIRROR_DIR}/main" &> /dev/null
+    # (These may need to be sorted)
+    output=\
+$'^refs/heads/master
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:303]: Why is there a $ here?

File: tests/int/test-git_mirror.bsh [hunk 7]
```diff
@@ -267,13 +288,47 @@ begin_test "Part 5 - Cloning from mirror"
...
+refs/heads/to_force1
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/master
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force1
+refs/tags/original_commit$'
+    [[ "$(git for-each-ref --format='%(refname)')" =~ ${output} ]] || false
+  popd &> /dev/null
+  pushd "${MIRROR_DIR}/recipes" &> /dev/null
+    output=\
+$'^refs/heads/master
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:312]: Why is there a `$` here?

File: tests/int/test-git_mirror.bsh [hunk 8]
```diff
@@ -267,13 +288,47 @@ begin_test "Part 5 - Cloning from mirror"
...
+  # git 1.8.3 does not support git -C
+  pushd "${MIRROR_DIR}/main" &> /dev/null
+    # (These may need to be sorted)
+    output=\
+$'^refs/heads/master
+refs/heads/to_force1
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/master
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force1
+refs/tags/original_commit$'
+    [[ "$(git for-each-ref --format='%(refname)')" =~ ${output} ]] || false
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:308]:
```suggestion
    [[ $(git for-each-ref --format='%(refname)') =~ ${output} ]] || false
```

File: tests/int/test-git_mirror.bsh [hunk 9]
```diff
@@ -267,13 +288,47 @@ begin_test "Part 5 - Cloning from mirror"
...
+  popd &> /dev/null
+  pushd "${MIRROR_DIR}/recipes" &> /dev/null
+    output=\
+$'^refs/heads/master
+refs/heads/to_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/master
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force2$'
+    [[ "$(git for-each-ref --format='%(refname)')" =~ ${output} ]] || false
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:316]:
```suggestion
    [[ $(git for-each-ref --format='%(refname)') =~ ${output} ]] || false
```

File: tests/int/test-git_mirror.bsh [hunk 10]
```diff
@@ -353,17 +435,63 @@ begin_test "Part 6 - Update a repo and the prepped mirror"
...
+  # Verify the refs are re-tagged by git_tag_main (still with the default regex)
+  #
+  # I could do these tests in the clone instead (the refs would be listed under
+  # refs/remotes and there would be a separate set of refs/heads; also, there
+  # are cases where certain tags are not fetched without the
+  # git fetch --tags flag)
+  pushd "${MIRROR_DIR}/main" &> /dev/null
+    # (These may need to be sorted)
+    output=\
+$'^refs/heads/a_branch
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:464]: `$`?

File: tests/int/test-git_mirror.bsh [hunk 11]
```diff
@@ -353,17 +435,63 @@ begin_test "Part 6 - Update a repo and the prepped mirror"
...
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force1
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/a_branch
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/master
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force1
+refs/tags/original_commit$'
+    [[ "$(git for-each-ref --format='%(refname)')" =~ ${output} ]] || false
+  popd &> /dev/null
+  pushd "${MIRROR_DIR}/recipes" &> /dev/null
+    output=\
+$'^refs/heads/master
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:478]: `$`?

File: tests/int/test-git_mirror.bsh [hunk 12]
```diff
@@ -353,17 +435,63 @@ begin_test "Part 6 - Update a repo and the prepped mirror"
...
+refs/heads/to_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/master
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force2$'
+    [[ "$(git for-each-ref --format='%(refname)')" =~ ${output} ]] || false
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:483]:
```suggestion
    [[ $(git for-each-ref --format='%(refname)') =~ ${output} ]] || false
```

File: tests/int/test-git_mirror.bsh [hunk 13]
```diff
@@ -397,6 +527,412 @@ begin_test "Part 6 - Update a repo and the prepped mirror"
...
+  # but I can't find a way to do this with git show-ref or git rev-parse)
+  #
+  # I could do these tests in the clone instead (the refs would be listed under
+  # refs/remotes and there would be a separate set of refs/heads; also, there
+  # are cases where certain tags are not fetched without the
+  # git fetch --tags flag)
+  pushd "${MIRROR_DIR}/main" &> /dev/null
+    # (These may need to be sorted)
+    output=\
+$'^refs/heads/a_branch
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:734]: `$`?

File: tests/int/test-git_mirror.bsh [hunk 14]
```diff
@@ -397,6 +527,412 @@ begin_test "Part 6 - Update a repo and the prepped mirror"
...
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/master
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force1
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/a_branch
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/master
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force1
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_can_force1
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_no_force1
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_to_force1
+refs/tags/original_commit$'
+    [[ "$(git for-each-ref --format='%(refname)')" =~ ${output} ]] || false
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:751]:
```suggestion
    [[ $(git for-each-ref --format='%(refname)') =~ ${output} ]] || false
```

File: tests/int/test-git_mirror.bsh [hunk 15]
```diff
@@ -397,6 +527,412 @@ begin_test "Part 6 - Update a repo and the prepped mirror"
...
+  popd &> /dev/null
+  pushd "${MIRROR_DIR}/recipes" &> /dev/null
+    output=\
+$'^refs/heads/master
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:755]: `$`

File: tests/int/test-git_mirror.bsh [hunk 16]
```diff
@@ -397,6 +527,412 @@ begin_test "Part 6 - Update a repo and the prepped mirror"
...
+refs/heads/to_tag_to_push_can_force2
+refs/heads/to_tag_to_push_no_force2
+refs/heads/to_tag_to_push_to_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/master
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_can_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_no_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_to_force2$'
+    [[ "$(git for-each-ref --format='%(refname)')" =~ ${output} ]] || false
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:767]:
```suggestion
    [[ $(git for-each-ref --format='%(refname)') =~ ${output} ]] || false
```

File: tests/int/test-git_mirror.bsh [hunk 17]
```diff
@@ -397,6 +527,412 @@ begin_test "Part 6 - Update a repo and the prepped mirror"
...
+  # are cases where certain tags are not fetched without the
+  # git fetch --tags flag)
+  pushd "${MIRROR_DIR}/main" &> /dev/null
+    # NOTE While to_tag_to_push_no_force1 is not force pushed, the tag to the
+    # updated commit is pushed:
+    #     just_git_mirror/{date}/heads/to_tag_to_push_no_force1
+    # RE While a little unexpected, this is not obviously wrong
+    # (These may need to be sorted)
+    output=\
+$'^refs/heads/a_branch
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:875]: `$`?

File: tests/int/test-git_mirror.bsh [hunk 18]
```diff
@@ -397,6 +527,412 @@ begin_test "Part 6 - Update a repo and the prepped mirror"
...
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_no_force1
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_to_force1
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_no_force1
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_to_force1
+refs/tags/original_commit$'
+    [[ "$(git for-each-ref --format='%(refname)')" =~ ${output} ]] || false
+  popd &> /dev/null
+  pushd "${MIRROR_DIR}/recipes" &> /dev/null
+    output=\
+$'refs/heads/master
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:898]: `$`?

File: tests/int/test-git_mirror.bsh [hunk 19]
```diff
@@ -397,6 +527,412 @@ begin_test "Part 6 - Update a repo and the prepped mirror"
...
+refs/heads/to_tag_to_push_to_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/master
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_can_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_no_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_to_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_no_force2
+refs/tags/just_git_mirror/[-.0-9TZ]+/heads/to_tag_to_push_to_force2$'
+    [[ "$(git for-each-ref --format='%(refname)')" =~ ${output} ]] || false
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:912]:
```suggestion
    [[ $(git for-each-ref --format='%(refname)') =~ ${output} ]] || false
```

**Review by andyneff — CHANGES_REQUESTED** [review]

File: tests/int/test-git_mirror.bsh (hunk 6 above)
Comment (scott-vsi) [reply tests/int/test-git_mirror.bsh:303]:
```suggestion
'^refs/heads/master
```

File: tests/int/test-git_mirror.bsh (hunk 7 above)
Comment (scott-vsi) [reply tests/int/test-git_mirror.bsh:312]:
```suggestion
'^refs/heads/master
```

File: tests/int/test-git_mirror.bsh (hunk 10 above)
Comment (scott-vsi) [reply tests/int/test-git_mirror.bsh:464]:
```suggestion
'^refs/heads/a_branch
```

File: tests/int/test-git_mirror.bsh (hunk 11 above)
Comment (scott-vsi) [reply tests/int/test-git_mirror.bsh:478]:
```suggestion
'^refs/heads/master
```

File: linux/git_mirror (hunk 4 above)
Comment (scott-vsi) [reply linux/git_mirror:924]: The users prompt was being changed when the call to the function ended

File: linux/git_mirror (hunk 1 above)
Comment (scott-vsi) [reply linux/git_mirror:865]: I actually do not want that behavior

File: linux/git_mirror (hunk 5 above)
Comment (scott-vsi) [reply linux/git_mirror:1008]: True, but git also says this

> "It’s generally recommended that you create annotated tags so you can have all this information; but if you want a temporary tag or for some reason don’t want to keep the other information, lightweight tags are available too.


https://git-scm.com/book/en/v2/Git-Basics-Tagging

File: tests/int/test-git_mirror.bsh [hunk 20]
```diff
@@ -686,14 +612,11 @@ begin_test "Part 7 - Rewrite a ref and mirror"
   popd &> /dev/null
 
   # Push to mirror
-  # NOTE These regex's are matching long-form refs
-  GIT_MIRROR_TAG_REFS_REGEX='.+to_tag_.+' \
-      GIT_MIRROR_PUSH_REFS_REGEX='.+_push.+' \
-      GIT_MIRROR_SKIP_REFS_REGEX='refs/heads/to_tag_skip_.+' \
+  # NOTE These regex's are matching long-form refs (this regex is using a
+  # negative look-ahead)
+  GIT_MIRROR_TAG_REFS_REGEX='^(?!no_).+$' \
```
Comment (andyneff) [inline tests/int/test-git_mirror.bsh:617]: Bash doesn't support look ahead

```
$ x="yolo yololo"
$ p='(?!lo)'
$ echo 'yolo yololo' | perl -wpe 's/yolo(?!lo)/YOLO/'
YOLO yololo
$ [[ ${x} =~ ${p} ]]
$ declare -p | grep BASH_RE
13:declare -ar BASH_REMATCH=()
```

Comment (scott-vsi) [reply tests/int/test-git_mirror.bsh:617]: So it would seem. It is odd that the tests all happily passed

File: linux/git_mirror [hunk 21]
```diff
@@ -852,14 +858,16 @@ function git_push_main()
...
           # Spaces are not allowed in a refname
           local refs=($("${GIT}" for-each-ref --format='%(refname)' | \
               grep -v -e '^refs/remotes' -e '^refs/stash' || :))
-          # xargs tries all args before returning an error code, only short-
-          # circuiting if the command exits with a value 255, which is not the
-          # case if pushing a branch fails
-          # from xarg's man page:
+          # xargs tries all args before returning an error code, as desired.
+          # it only short-circuits if the command exits with a value 255, which
+          # is not the case if pushing a branch fails. Per xarg's man page:
```
Comment (andyneff) [inline linux/git_mirror:870]:
```suggestion
          # It only short-circuits if the command exits with a value 255, which
          # luckily is not the case if pushing a branch fails. Per xarg's man page:
```

**Review by andyneff — APPROVED** [review]
